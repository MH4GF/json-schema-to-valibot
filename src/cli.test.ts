import { spawnSync } from 'node:child_process'
import { readFileSync, readdirSync } from 'node:fs'
import { mkdirSync } from 'node:fs'
import { dirname, join } from 'node:path'
import { describe, expect, it } from 'vitest'

describe('CLI', () => {
  it('should display help message', () => {
    const { stdout } = spawnSync('node', ['src/cli.ts', '--help'], {
      encoding: 'utf8',
    })

    expect(stdout).toMatchInlineSnapshot(`
      "Usage: json-schema-to-valibot [options]

      Convert JSON Schema to Valibot schema

      Options:
        -V, --version        output the version number
        -i, --input <path>   JSON or a source file path
        -o, --output <path>  A file path to write to
        -n, --name <name>    The name of the schema in the output
        -m, --module <type>  Module syntax: 'esm', 'cjs' or 'none' (default: "esm")
        -t, --type <name>    Export a named type along with the schema
        --no-import          Removes the import statement from the output
        -j, --with-jsdocs    Generate jsdocs off of the description property
        -h, --help           display help for command
      "
    `)
  })

  it('should run with input file only', () => {
    const { stderr } = spawnSync(
      'node',
      [
        '--no-warnings',
        'src/cli.ts',
        '-i',
        'src/test/JSON-Schema-Test-Suite/tests/latest/default.json',
      ],
      {
        encoding: 'utf8',
      },
    )

    expect(stderr).toBeFalsy()
  })

  it('should read from stdin', () => {
    const { stderr } = spawnSync('node', ['--no-warnings', 'src/cli.ts'], {
      input: '{"type": "string"}',
      encoding: 'utf8',
    })

    expect(stderr).toBeFalsy()
  })

  it('should write to output file', () => {
    const outputPath = 'src/test/output/default.js'
    mkdirSync(dirname(outputPath), { recursive: true })

    const { stderr } = spawnSync(
      'node',
      [
        '--no-warnings',
        'src/cli.ts',
        '-i',
        'src/test/JSON-Schema-Test-Suite/tests/latest/default.json',
        '--output',
        outputPath,
      ],
      {
        encoding: 'utf8',
      },
    )

    expect(stderr).toBeFalsy()

    const output = readFileSync(outputPath, 'utf8')
    expect(output).toMatchSnapshot()
  })

  it('should write to output file without import', () => {
    const outputPath = 'src/test/output/default-no-import.js'
    mkdirSync(dirname(outputPath), { recursive: true })

    const { stderr } = spawnSync(
      'node',
      [
        '--no-warnings',
        'src/cli.ts',
        '-i',
        'src/test/JSON-Schema-Test-Suite/tests/latest/default.json',
        '--output',
        outputPath,
        '--no-import',
      ],
      {
        encoding: 'utf8',
      },
    )

    expect(stderr).toBeFalsy()

    const output = readFileSync(outputPath, 'utf8')
    expect(output).toMatchSnapshot()
  })

  it('should write to output file with JSDoc comments', () => {
    const outputPath = 'src/test/output/default-with-jsdoc.js'
    mkdirSync(dirname(outputPath), { recursive: true })

    const { stderr } = spawnSync(
      'node',
      [
        '--no-warnings',
        'src/cli.ts',
        '-i',
        'src/test/JSON-Schema-Test-Suite/tests/latest/default.json',
        '--output',
        outputPath,
        '--with-jsdocs',
      ],
      {
        encoding: 'utf8',
      },
    )

    expect(stderr).toBeFalsy()

    const output = readFileSync(outputPath, 'utf8')
    expect(output).toMatchSnapshot()
  })

  it('should error when no input is provided', () => {
    const { stderr } = spawnSync('node', ['--no-warnings', 'src/cli.ts'], {
      encoding: 'utf8',
    })

    expect(stderr).toMatchInlineSnapshot(`
      "Input is required when no JSON or file path is piped
      "
    `)
  })

  // biome-ignore lint/suspicious/noSkippedTests: <explanation>
  it.skip('should error on invalid module syntax', () => {
    const { stderr } = spawnSync(
      'node',
      [
        '--no-warnings',
        'src/cli.ts',
        '-i',
        'src/test/JSON-Schema-Test-Suite/tests/latest/default.json',
        '-m',
        'invalid',
      ],
      {
        encoding: 'utf8',
      },
    )

    expect(stderr).toMatchInlineSnapshot(`
      "Invalid module syntax: invalid
      "
    `)
  })

  describe('JSON Schema Test Suite', () => {
    const testDir = 'src/test/JSON-Schema-Test-Suite/tests/latest'
    const testFiles = readdirSync(testDir).filter((file) => file.endsWith('.json'))

    for (const testFile of testFiles) {
      it(testFile, () => {
        const outputPath = `src/test/output/${testFile}.ts`
        mkdirSync(dirname(outputPath), { recursive: true })

        const { stderr } = spawnSync(
          'node',
          ['--no-warnings', 'src/cli.ts', '-i', join(testDir, testFile), '--output', outputPath],
          {
            encoding: 'utf8',
          },
        )

        expect(stderr).toBeFalsy()
        const output = readFileSync(outputPath, 'utf8')
        expect(output).toMatchSnapshot()
      })
    }
  })
})
