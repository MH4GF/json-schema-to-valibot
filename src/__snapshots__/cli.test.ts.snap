// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CLI > JSON Schema Test Suite > additionalProperties.json 1`] = `
"import * as v from "valibot";

export default v.any();"
`;

exports[`CLI > JSON Schema Test Suite > allOf.json 1`] = `
"import * as v from "valibot";

export default v.any();"
`;

exports[`CLI > JSON Schema Test Suite > anchor.json 1`] = `
"import * as v from "valibot";

export default v.any();"
`;

exports[`CLI > JSON Schema Test Suite > anyOf.json 1`] = `
"import * as v from "valibot";

export default v.any();"
`;

exports[`CLI > JSON Schema Test Suite > boolean_schema.json 1`] = `
"import * as v from "valibot";

export default v.any();"
`;

exports[`CLI > JSON Schema Test Suite > const.json 1`] = `
"import * as v from "valibot";

export default v.any();"
`;

exports[`CLI > JSON Schema Test Suite > contains.json 1`] = `
"import * as v from "valibot";

export default v.any();"
`;

exports[`CLI > JSON Schema Test Suite > content.json 1`] = `
"import * as v from "valibot";

export default v.any();"
`;

exports[`CLI > JSON Schema Test Suite > default.json 1`] = `
"import * as v from "valibot";

export default v.any();"
`;

exports[`CLI > JSON Schema Test Suite > defs.json 1`] = `
"import * as v from "valibot";

export default v.any();"
`;

exports[`CLI > JSON Schema Test Suite > dependentRequired.json 1`] = `
"import * as v from "valibot";

export default v.any();"
`;

exports[`CLI > JSON Schema Test Suite > dependentSchemas.json 1`] = `
"import * as v from "valibot";

export default v.any();"
`;

exports[`CLI > JSON Schema Test Suite > dynamicRef.json 1`] = `
"import * as v from "valibot";

export default v.any();"
`;

exports[`CLI > JSON Schema Test Suite > enum.json 1`] = `
"import * as v from "valibot";

export default v.any();"
`;

exports[`CLI > JSON Schema Test Suite > exclusiveMaximum.json 1`] = `
"import * as v from "valibot";

export default v.any();"
`;

exports[`CLI > JSON Schema Test Suite > exclusiveMinimum.json 1`] = `
"import * as v from "valibot";

export default v.any();"
`;

exports[`CLI > JSON Schema Test Suite > format.json 1`] = `
"import * as v from "valibot";

export default v.any();"
`;

exports[`CLI > JSON Schema Test Suite > if-then-else.json 1`] = `
"import * as v from "valibot";

export default v.any();"
`;

exports[`CLI > JSON Schema Test Suite > infinite-loop-detection.json 1`] = `
"import * as v from "valibot";

export default v.any();"
`;

exports[`CLI > JSON Schema Test Suite > items.json 1`] = `
"import * as v from "valibot";

export default v.any();"
`;

exports[`CLI > JSON Schema Test Suite > maxContains.json 1`] = `
"import * as v from "valibot";

export default v.any();"
`;

exports[`CLI > JSON Schema Test Suite > maxItems.json 1`] = `
"import * as v from "valibot";

export default v.any();"
`;

exports[`CLI > JSON Schema Test Suite > maxLength.json 1`] = `
"import * as v from "valibot";

export default v.any();"
`;

exports[`CLI > JSON Schema Test Suite > maxProperties.json 1`] = `
"import * as v from "valibot";

export default v.any();"
`;

exports[`CLI > JSON Schema Test Suite > maximum.json 1`] = `
"import * as v from "valibot";

export default v.any();"
`;

exports[`CLI > JSON Schema Test Suite > minContains.json 1`] = `
"import * as v from "valibot";

export default v.any();"
`;

exports[`CLI > JSON Schema Test Suite > minItems.json 1`] = `
"import * as v from "valibot";

export default v.any();"
`;

exports[`CLI > JSON Schema Test Suite > minLength.json 1`] = `
"import * as v from "valibot";

export default v.any();"
`;

exports[`CLI > JSON Schema Test Suite > minProperties.json 1`] = `
"import * as v from "valibot";

export default v.any();"
`;

exports[`CLI > JSON Schema Test Suite > minimum.json 1`] = `
"import * as v from "valibot";

export default v.any();"
`;

exports[`CLI > JSON Schema Test Suite > multipleOf.json 1`] = `
"import * as v from "valibot";

export default v.any();"
`;

exports[`CLI > JSON Schema Test Suite > not.json 1`] = `
"import * as v from "valibot";

export default v.any();"
`;

exports[`CLI > JSON Schema Test Suite > oneOf.json 1`] = `
"import * as v from "valibot";

export default v.any();"
`;

exports[`CLI > JSON Schema Test Suite > pattern.json 1`] = `
"import * as v from "valibot";

export default v.any();"
`;

exports[`CLI > JSON Schema Test Suite > patternProperties.json 1`] = `
"import * as v from "valibot";

export default v.any();"
`;

exports[`CLI > JSON Schema Test Suite > prefixItems.json 1`] = `
"import * as v from "valibot";

export default v.any();"
`;

exports[`CLI > JSON Schema Test Suite > properties.json 1`] = `
"import * as v from "valibot";

export default v.any();"
`;

exports[`CLI > JSON Schema Test Suite > propertyNames.json 1`] = `
"import * as v from "valibot";

export default v.any();"
`;

exports[`CLI > JSON Schema Test Suite > ref.json 1`] = `
"import * as v from "valibot";

export default v.any();"
`;

exports[`CLI > JSON Schema Test Suite > refRemote.json 1`] = `
"import * as v from "valibot";

export default v.any();"
`;

exports[`CLI > JSON Schema Test Suite > required.json 1`] = `
"import * as v from "valibot";

export default v.any();"
`;

exports[`CLI > JSON Schema Test Suite > type.json 1`] = `
"import * as v from "valibot";

export default v.any();"
`;

exports[`CLI > JSON Schema Test Suite > unevaluatedItems.json 1`] = `
"import * as v from "valibot";

export default v.any();"
`;

exports[`CLI > JSON Schema Test Suite > unevaluatedProperties.json 1`] = `
"import * as v from "valibot";

export default v.any();"
`;

exports[`CLI > JSON Schema Test Suite > uniqueItems.json 1`] = `
"import * as v from "valibot";

export default v.any();"
`;

exports[`CLI > JSON Schema Test Suite > vocabulary.json 1`] = `
"import * as v from "valibot";

export default v.any();"
`;

exports[`CLI > should write to output file 1`] = `
"import * as v from "valibot";

export default v.any();"
`;

exports[`CLI > should write to output file with JSDoc comments 1`] = `
"import * as v from "valibot";

export default v.any();"
`;

exports[`CLI > should write to output file without import 1`] = `"export default v.any();"`;
